recipes/full_finetune_single_device.py

```diff
@@ -418,6 +425,58 @@ def _setup_optimizer(
             log.info("Optimizer is initialized.")
             return optimizer
 
+    def _setup_lr_scheduler(
+        self,
+        cfg_lr_scheduler: DictConfig,
+        num_training_steps: int,
+        last_epoch: int,
+    ) -> Optional[Optimizer]:
+        """
+        Set up the learning rate scheduler based on the provided configuration.
+        It handles both standard optimization and optimizer-in-backward cases, and supports
+        schedulers from both torchtune.modules and torch.optim.
+
+        Args:
+            cfg_lr_scheduler (DictConfig): The learning rate scheduler configuration.
+            num_training_steps (int): The total number of training steps.
+            last_epoch (int): The index of the last epoch.
+
+        Returns:
+            lr_scheduler (Optional[Optimizer]): The learning rate scheduler.
+        """
+        if self._optimizer_in_bwd:
+            # Use the first optimizer from the wrapper to represent the learning rate
+            optimizer = next(iter(self._optim_ckpt_wrapper.optim_map.values()))
+        else:
+            # Standard case: use the single optimizer
+            optimizer = self._optimizer
+
+        # Check if the lr_scheduler component is from torchtune.modules or torch.optim
+        component = cfg_lr_scheduler.get("_component_", None)
+
+        # Conditionally instantiate the scheduler based on the component to support torch.optim
+        if "torchtune.modules" in component:
+            # Instantiate the learning rate scheduler
+            lr_scheduler = config.instantiate(
+                cfg_lr_scheduler,
+                optimizer,
+                num_training_steps=num_training_steps,
+                last_epoch=last_epoch,
+            )
+        else:
+            lr_scheduler = config.instantiate(
+                cfg_lr_scheduler,
+                optimizer,
+                last_epoch=last_epoch,
+            )
```

Comment:
Ah is this a consequence of using `ConstantLR` in the test? I think it's a bit confusing to condition on whether the component is coming from torchtune.modules or not. In general there will be things from torch.optim that don't match the signature in the else statement, and there will probably also (in the future) be things from torchtune.modules that don't match the signature in the if statement.

In that case (and sorry for changing my mind on this), is there any reason we can't just make `lr_scheduler` optional? Then if it's not passed we just continue to use the usual constant learning rate defined in the optimizer and set `self._lr_scheduler=None` in the recipe (and in the training loop check that it's not None prior to stepping). This will make the recipe test a non-issue and keep the setup clean. 

Separately though it's clear that our LR scheduler instantiation is not flexible enough. We should figure out how to better support schedulers from torch.optim moving forward (though that's a problem for another PR).

Reply:
Yes. I was thinking to make it more flexible also and support other types of torch.optim.lr_scheduler. I find that torchtune.modules.lr_scheduler only has get_cosine_schedule_with_warmup and num_training_steps input would not be compatible with other torch.optim.lr_scheduler. 

I can check and work with lr_scheduleer as None by the weekend. It is mainly due to the testing config that it always would input num_training_steps even when there is no args for that.

Reply:
Updated `self._lr_scheduler=None`
Also I modified the testing. I still use the cosine instead of constant but effectively change it equivalent to constant by setting `num_cycles=0`. It should be a bit more elegant imo